### ISSUE TITLE
Ampersand encoding in UrlBuilder

### ISSUE THREAD

**Issue by WimVergouwe** [mikeedwards83/Glass.Mapper#197]
Encoding of & in UrlBuilder breaks media url protection of Sitecore.
I guess UrlBuilder shouldn't encode ampersands.

Sample test:
```C#
using System;
using Glass.Mapper;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Sitecore.Resources.Media;

namespace Dlw.ScBase.Website.IntegrationTest.Infrastructure.GlassORM
{
    /// <summary>
    ///     Issue: Server resizing of images rendered with Glass does not always work.
    ///     When: URL / query string contains a query string parameter
    ///     * that is not a protected media query parameter
    ///     * and appears as the first parameter of the query string.
    ///     (e.g. "?custom=value&h=200&w=200", but not "?h=200&custom=value&w=200")
    ///     Why:
    ///     Glass (GlassHtml.RenderImage) passes the img src as a URL encoded string to HashingUtils.ProtectAssetUrl.
    ///     HashingUtils.ProtectAssetUrl only protects (adding a hash query string parameter) if the string contains
    ///     query string parameters to protect (configured in Sitecore.Media.RequestProtection protectedMediaQueryParameters).
    ///     It does this by checking if the string contains "&paramname" or "?paramname" (HashingUtils.IsSafeUrl).
    ///     This first check always fails for encoded strings: "&amp;paramname" does not match anymore.
    ///     (e.g. "?custom=value&h=200&w=200" became "?custom=value&amp;h=200&amp;w=200", so "&h" and "?h" don't match)
    ///     If the URL is not protected with a hash, resizing is not applied.
    ///     If the custom, not protected, query string parameter is not the first parameter,
    ///     everything works fine because "?paramname" is not different when URL encoded:
    ///     "?h" still matches for "?h=200&amp;custom=value&amp;w=200"
    ///
    ///     Atm there is no real code fix as impacted code can't be changed that easily.
    ///     Work around is to add URL parameters to the protectedMediaQueryParameters.
    /// </summary>
    [TestClass]
    public class ProtectAssetUrlsTests : AbstractSitecoreTest
    {
        /// <summary>
        ///     GlassHtml.RenderImage uses UrlBuilder.ToString, which encodes the URL.
        ///     Encoded URLs give issues when protecting.
        /// </summary>
        [TestMethod]
        public void CanProtectedAssetUrlBuiltByGlass()
        {
            // no custom parameter
            var assetUrl = HashingUtils.ProtectAssetUrl("/~/media/image.ashx?bc=white&h=250&w=250");
            Assert.IsTrue(IsProtected(assetUrl), assetUrl);
            var urlBuilder = new UrlBuilder("/~/media/image.ashx?bc=white&h=250&w=250");
            assetUrl = HashingUtils.ProtectAssetUrl(urlBuilder.ToString());
            Assert.IsTrue(IsProtected(assetUrl), assetUrl);
           
            // custom query string parameter in the middle => OK
            assetUrl = HashingUtils.ProtectAssetUrl("/~/media/image.ashx?bc=white&type=thumbnail&h=250&w=250");
            Assert.IsTrue(IsProtected(assetUrl), assetUrl);
            urlBuilder = new UrlBuilder("/~/media/image.ashx?bc=white&type=thumbnail&h=250&w=250");
            assetUrl = HashingUtils.ProtectAssetUrl(urlBuilder.ToString());
            Assert.IsTrue(IsProtected(assetUrl), assetUrl);
            
            // custom query string parameter at the front => FAIL when encoded
            assetUrl = HashingUtils.ProtectAssetUrl("/~/media/image.ashx?type=thumbnail&bc=white&h=250&w=250");
            Assert.IsTrue(IsProtected(assetUrl), assetUrl); // OK

            urlBuilder = new UrlBuilder("/~/media/image.ashx?type=thumbnail&bc=white&h=250&w=250");
            assetUrl = HashingUtils.ProtectAssetUrl(urlBuilder.ToString());
            Assert.IsTrue(IsProtected(assetUrl), assetUrl); // FAIL becasue ToString encodes
        }
        private static bool IsProtected(string assetUrl)
        {
            return assetUrl.IndexOf("hash=", StringComparison.OrdinalIgnoreCase) >= 0;
        }
    }
}

**Comment by mikeedwards83** [comment]
Hi

Glass automatically adds the hash to the end of the URL if you use RenderImage:

       @RenderImage(x => x.Image, isEditable: true)

Will output the following:

    <img src="/sitecore/shell/-/media/Experience-Explorer/Presets/Unknown-128x128.ashx?h=128&amp;w=128&amp;la=en&amp;hash=A023B2E65106AC64654566D9294071C488C003A0" alt="Unknown">

**Comment by Elyseum** [comment]
Hi,
if you add a custom parameter which is not a sitecore protected media query parameter to the RenderImage method and it shows up as the first parameter of the query string, it will not work.
e.g. "?notprotected=avalue&h=123&w=123..." instead of "?h=123&notprotected&w=123..."

Before applying the hash, Sitecore checks wether the query string has protected parameters (HashingUtils.ProtectAssetUrl). It does this by checking for "&name" and "?name"; with a too simple string.contains.  Since Glass applies encoding, the match for "&name" no longer works (it became &amp;). So if the first parameter ("?notprotected=") is in fact not protected, the matching does not work (other protected ones are "&amp;") and no hash is added to the rendering. Your example works because the first parameter is ?h, which is outputted the same, encoded or not.

**Comment by mikeedwards83** [comment]
I have pushed a fix to the nightly build 4.0.5.98, could you try this?

**Comment by Elyseum** [comment]
Hi,

I tried given unit test with version 4.0.5.98-beta and it seems to work fine now :). Thx for the fix.

**Comment by mikeedwards83** [comment]
This has been released in 4.0.6.55
